### PR TITLE
Allow users to specify how long a message should stay on screen

### LINKED ISSUE #15: Configuration for deleting messages after an amount of time has elapsed
## Need

Some users would like messages to go away after some amount of time. This could be handled within a given theme using CSS animation, but this depends on the theme supporting the functionality. Besides, even if a theme were to support deleting messages after some amount of time, that duration would likely be hardcoded.

**showmy.chat should provide a query parameter through which users can specify a duration after which the message will be deleted.** If this query parameter is not provided, then the default behavior should be to not delete any messages — this way, users strictly opt in to messages deleting.

## Suggested API

- Configuration should be strictly optional. If query parameter is not provided, messages are not deleted after any interval.
    - In homepage URL builder, if user does not opt in to deleting after a given interval, don't add the query parameter to the built URL at all
- Query parameter could be named something like `clearMessageAfter` or `deleteMessageAfter`. Open to other names if folks have preferences.
- Query parameter value should be a number, representing the interval duration before the message is deleted.
    - I'm undecided on whether this should represent number of milliseconds (expected integer) or number of seconds (expected integer or float). I'm open to stronger opinions than mine.
- Deletion should be handled in the JavaScript, and deleted messages should be completely deleted, rather than hidden. This is in part because some folks may be "delete-after" feature to reduce the number of DOM elements on the page for perf reasons
- Users should (optionally) be able to combine this query parameter with the `showLatestMessages` param from #14 if they so choose — these parameters should be treated as independent
- Query parameter data should be provided to the scripts using the same pattern as #14 

## Steps

- [ ] Decide on a name for the query parameter
- [ ] Decide whether we expect milliseconds (integer) or seconds (integer or float) as the query parameter value
- [ ] Ensure Eleventy Serverless query parameter data can be accessed inside `handleChat` script
- [ ] Ensure if the `clearMessageAfter` (or whichever other name we land on) query parameter has been provided, and is a valid number, scripts delete messages after that interval
- [ ] Add configuration to homepage URL builder
- [ ] Remove any hardcoded CSS implementation in themes

### PR DESCRIPTION
**Pull request by BenDMyers**
Resolves #15 — a fairly old issue.

This pull request introduces a `clearMessageAfter` query parameter for the overlays, which accepts a positive integer. This is treated as number of milliseconds after which a message should be removed rom the overlay (assuming it hasn't already been deleted by, for instance, mods)

Messages that are about to be deleted are now given the attribute `data-twitch-message-status="deleting"`, as a style hook for themes to begin animating/transitioning for the message's departure. Because there's now a _lot_ going on with message deletion, the approach has been standardized with a new `removeMessage` helper function.

An example outbound animation can be seen in the `default` and `animalese` themes.

### CODE REVIEW

**Comment by NintenZone** [comment]
Hey Ben! After a brief look, this looks great! Very essential feature.

**Comment by KenAKAFrosty** [comment]
Looks great, very readable! The inclusion of the deleting attribute will be great for adding that extra pizzazz to themes. 

One thing to consider perhaps is to use, or at least expose to the end user, the more familiar unit of seconds instead of ms. We deal with milliseconds all the time but many casual users probably don't 😆

Maybe not even necessarily in the query param itself, but just in the configuration UI on the home page?

**Comment by BenDMyers** [comment]
@KenAKAFrosty Great feedback! Let me give that a try!

File: src/index.html
Comment (NickyMeuleman) [inline src/index.html:147]:
```suggestion
			if (chosenValue == defaultValue) {
```
Now that the default for `clearMessageAfter` is an empty string like `hideMessagesFrom`, the check that's in place should cover the case where this input isn't touched during urlbuilding.

I like the renaming to  `defaultValue`, it makes the intent clearer!

File: src/scripts/handleChat.mjs
Comment (NickyMeuleman) [reply src/scripts/handleChat.mjs:242]: Ty!

File: src/index.html
Comment (NickyMeuleman) [inline src/index.html:56]:
```suggestion
			placeholder="{{defaults.clearMessageAfter}}" step="0.1" min="0"/>
```
Preventing negative numbers from being entered clientside.

File: src/scripts/utilities.mjs
Comment (NickyMeuleman) [inline src/scripts/utilities.mjs:89]: This feels a bit weird for themes without an animation.
I'd expect messages to disappear after the time I set in the urlbuilder on the homepage.
For themes with an animation: the animation starts at the time I set.
For themes without: they stay onscreen for the time I set + 1.2s

**Review by NickyMeuleman — CHANGES_REQUESTED** [review]
Nicely done! You put a lot of thought into this.
Works well, handles corner cases (the question I asked about trying to delete messages that were already gone).

My only additional question is about that hardcoded `1200`.

File: src/index.html
Comment (BenDMyers) [reply src/index.html]: I merged in this change, and realized a case where this `chosenValue === ''` check was needed: the `showLatestMessages` field. The default value for that is `100`, but that's now being represented as a placeholder, and not an initial value. As a result, leaving the field blank (`''`) is not going to be equal to the default (`100`). I'm adding this check back in — but thank you for the suggestion!

File: src/scripts/utilities.mjs
Comment (BenDMyers) [reply src/scripts/utilities.mjs:89]: I agree, this is an odd one for sure - but I'm not sure how to do it better. I suppose we could find a way to track "Was an animation or a transition started by applying the `[data-twitch-message-display-status="deleting"]` style hook?"

In practice, I think this weirdness will be mitigated by two things:

1. More themes having exit animations _(assuming those animations are less than 1.2 seconds long, which, tbf, they probably should be)_
2. Folks will likely choose longer message lifespans _(That extra 1.2 seconds matters less and less when your message lifespan is, say, 20 seconds than when it's 5 seconds)_.

That said, if we know of a way to do away with that 1.2 sec delay, I'm definitely open to it!

File: src/index.html
Comment (NickyMeuleman) [reply src/index.html:147]: Ah good catch. It changed to placeholder to keep consistency with the "Show Commands" input?
To make sure my understanding is correct: now the defaults are used in the serverless function if nothing is provided for that option in the URL builder. The defaults are only shown in the URLbuilder as guides for "this is a possible thing that goes here".

In that case, can we consolidate all default checks? Changing them to "does this input option contain anything"?
Unsure, will test by console.logging later.

With the `=== ''` check, lines 140-143 will never be hit, so those can go too.
https://github.com/BenDMyers/showmy.chat/pull/91/files#diff-ce434ef5cc161194b450135a3586eb89357fc270e8aaa8e336db5a04e8d52fbeR140-R143

File: src/scripts/utilities.mjs
Comment (BenDMyers) [reply src/scripts/utilities.mjs:89]: @NickyMeuleman — I've been looking into this, and it looks like we could deduce whether there is an active transition or animation on the element.

Active transition: `!!messageToDelete.style.transition`
Active animation: Using the [`getAnimations`](https://developer.mozilla.org/en-US/docs/Web/API/Element/getAnimations) method which dropped in Chrome 84.

Using either of these seems like it could be fraught with heuristics, but it might be worth investigating.

Chrome 84 is more on the table now that folks who use OBS for Twitch are forced to update shortly, but I'd still like to hold off for a _bit_ before we start using `>75` APIs.

My proposal is that we ship this as it is, but file a new issue for revisiting this heuristic here a few months down the line. Thoughts?

Comment (NickyMeuleman) [reply src/scripts/utilities.mjs:89]: I agree, it's definitely a nice to have, but too soon at the moment. (TIL about that API, thanks for showing it to me, seems awesome!)
Keep as is, the confusion will be minimal (and only be there on themes that don't use animations, and probably only if the delay is short enough to notice).

(I didn't respond previously because I don't have a suggestion on how to handle this)

Comment (BenDMyers) [reply src/scripts/utilities.mjs:89]: I'm going to resolve this for now, and I'll create the issue once this PR is merged in!

Comment (NickyMeuleman) [reply src/scripts/utilities.mjs:89]: revisit that == '' check please. The question wheter that check serves the same purpose as the 
```
if (key == 'showCommands' && chosenValue === '') {
			// the showCommands option remained unchanged and should use the default
			chosenValue = defaultQueryParameters.showCommands;
}
```
Block.
Other than that minor detail, this looks great (I'd like to fold in that logic to the new check as well if that doesn't cause additional bugs I didn't think of). In other words: can we also delete that whole section now?

File: src/index.html
Comment (BenDMyers) [reply src/index.html:147]: Ah, great catch! I had totally missed that this negated the need for the `if (key == 'showCommands' && chosenValue === '')` check - I can go and remove that!

Comment (BenDMyers) [reply src/index.html:147]: I think the question of "if someone manually puts a value that's the same as the default, should that be explicitly passed in the URL" is a good one, and one that I'm not sure I have an answer for!

Allowing folks to more explicitly set values equal to the defaults would mean that if, for some reason, we decided to change a default (a move that should definitely be done rarely, but which I don't think is entirely out of the question), those users' overlays don't change from underneath them. That's not, altogether, a bad idea.

Maybe this check is best handled as:

```suggestion
if (chosenValue === '') {
```
?

Comment (NickyMeuleman) [reply src/index.html:143]: Yep, that should catch the inputs that haven't been touched, and include everything that was explicitly set!
Let's use this.

**Review by NickyMeuleman — APPROVED** [review]

**Pull request merged**
